// contentcube/bootstrap.js
var FileSystem = require("contentcube/filesystem");
var Path = require("path");
var Step = require("step");

exports.modules = function(modulesPath, callback) {
	Step(
		function getModules() {
			FileSystem.readdir(modulesPath, this);
		},
		function fileListReceived(error, fileNames) {
			if (error) throw error;

			var group = this.group();
			fileNames.forEach(function(filename) {
				var moduleDirectoryPath = Path.join(modulesPath, filename);
				FileSystem.stat(moduleDirectoryPath, group());
			});
		},
		function fileStatReceived(error, fileStats) {
			if (error) throw error;
			var callback = this;
			var bootstrapFilePaths = [];

			function checkForBootstrapExistence() {
				if (fileStats.length <= 0) {
					return callback(null, bootstrapFilePaths);
				}

				var file = fileStats.pop();
				if (!file.isDirectory()) {
					checkForBootstrapExistence();
				}
				var modulePath = file.filepath;

				var bootstrapFilePath = Path.join(modulePath, 'bootstrap.js');
				Path.exists(bootstrapFilePath, function(exists) {
					if (exists) {
						bootstrapFilePaths.push(bootstrapFilePath);
					}

					checkForBootstrapExistence();
				});
			}

			checkForBootstrapExistence();
		},
		function executeBootstraps(error, bootstrapPaths) {
			if (error) throw error;

			bootstrapPaths.forEach(function(bootstrapPath) {
				var bootstrap = require(bootstrapPath);
				bootstrap();
				console.log("initialized: " + bootstrapPath);
			});
			
			callback(error);
		}	
	);
};
