// contentcube/helpers/broker.js
var ejs = require("ejs");
var fileSystem = require("fs");
var path = require("path");

module.exports = function(request) {
	var instance = {};
	
	instance.url = {
		current: function() {
			var args = Array.prototype.slice.call(arguments);
			args.unshift(request.url);
			var uri = path.join.apply(path, args);
			return uri;
		},
		baseUrl: function() {
			var args = Array.prototype.slice.call(arguments);
			return global.application.config.get('system.baseUrl') + path.join.apply(path, args);
		},
		build: function(route) {
			var uri = application.config.get('system.baseUrl');
			
			if (application.config.get('system.modulesEnabled')) {
				uri = path.join(uri, route.module || application.config.get('routing.defaultModuleName'));
			}
			
			uri = path.join(uri, route.controller || application.config.get('routing.defaultControllerName'));
			uri = path.join(uri, route.action || application.config.get('routing.defaultActionName'));
			
			if (route.params) {
				uri = path.join(uri, route.params.join("/")); 
			}
			
			return uri;
		}
	};
	
	/**
	 * Getter / Setter for placeholders.
	 * @param {String} key
	 * @return {Object}
	 */
	instance.placeholder = function(key) {	
		var placeholder = request.placeholder(key);
		
		return {
			get: function(key, defaultValue) {
				if (!key) {
					return placeholder;
				}
				
				return placeholder[key] || defaultValue;
			},
			set: function(key, value) {
				placeholder[key] = value;
			}
		};
	};
	
	/**
	 * Render and a view, notice this is a blocking call.
	 * @param {String} viewName
	 * @param {Object} viewData
	 * @return {String}
	 */
	instance.renderView = function(viewName, viewData) {
		var pathToView = '/application/views/' + viewName;
		if (global.application.config.get('system.modulesEnabled', false)) {
			pathToView = '/application/modules/' + request.moduleName + '/views/' + viewName;
		}
		
		var filepath = path.join(global.application.config.get('system.localPath'), path.normalize(pathToView));
		
		var fileContent = fileSystem.readFileSync(filepath, 'UTF8');
		viewData._helpers = instance;
		var html = ejs.render(fileContent, {locals: viewData});
		
		return html;
	};
	
	/**
	 * Wrapper around a placeholder to store data related to rendering the layout.
	 */
	instance.layout = (function() {
		var placeholder = instance.placeholder('_layout');
			placeholder.scripts = [];
			placeholder.styles = [];
		
		var layoutHelper = {
			setTitle: function(title) {
				placeholder.set('title', title);
			},
			getTitle: function() {
				return placeholder.get('title', '');
			},
			addScript: function(url) {
				if (placeholder.scripts.indexOf(url) == -1) {
					placeholder.scripts.push(url);
				}
			},
			getScripts: function() {
				var html = [];
				placeholder.scripts.forEach(function(url) {
					html.push('<script type="text/javascript" src="' + url + '"></script>');
				});
				
				return html.join("\n");
			},
			addStyle: function(url) {
				if (placeholder.styles.indexOf(url) == -1) {
					placeholder.styles.push(url);
				}
			},
			getStyles: function() {
				var html = [];
				placeholder.styles.forEach(function(url) {
					html.push('<link rel="stylesheet" type="text/css" href="' + url + '" media="all" />');
				});
				
				return html.join("\n");
			}
		};
		
		return layoutHelper;
	})();
	
	return instance;
};
